src/polycap-description.c

```diff
@@ -0,0 +1,455 @@
+{
+	int i;
+	polycap_description *description;
+
+	//allocate some memory
+	description = malloc(sizeof(polycap_description));
+	if(description == NULL){
+		printf("Could not allocate description memory.\n");
+		exit(1);
+	}
+	description->iz = malloc(sizeof(int)*nelem);
+	if(description->iz == NULL){
+		printf("Could not allocate description->iz memory.\n");
+		exit(1);
+	}
+	description->wi = malloc(sizeof(double)*nelem);
+	if(description->wi == NULL){
+		printf("Could not allocate description->wi memory.\n");
+		exit(1);
+	}
+
+	//copy data into description structure
+	description->sig_rough = sig_rough;
+	description->sig_wave = sig_wave;
+	description->corr_length = corr_length;
+	description->n_cap = n_cap;
+	description->d_source = d_source;
+//	description->d_screen = d_screen;
+	description->src_x = src_x;
+	description->src_y = src_y;
+	description->src_sigx = src_sigx;
+	description->src_sigy = src_sigy;
+	description->src_shiftx = src_shiftx;
+	description->src_shifty = src_shifty;
+	description->nelem = nelem;
+	description->density = density;
+	for(i=0; i<description->nelem; i++){
+		description->iz[i] = iz[i];
+		description->wi[i] = wi[i]; //assumes weights are already provided as fractions (not percentages)
+	}
+
+	// Check whether weights add to 1
+	polycap_description_check_weight(description->nelem, description->wi);
+
+	//allocate profile memory
+	description->profile = malloc(sizeof(polycap_profile));
+	if(description->profile == NULL){
+		printf("Could not allocate description->profile memory.\n");
+		exit(1);
+	}
+	description->profile->z = malloc(sizeof(double)*(profile->nmax+1));
+	if(description->profile->z == NULL){
+		printf("Could not allocate description->profile->z memory.\n");
+		exit(1);
+	}
+	description->profile->cap = malloc(sizeof(double)*(profile->nmax+1));
+	if(description->profile->cap == NULL){
+		printf("Could not allocate description->profile->cap memory.\n");
+		exit(1);
+	}
+	description->profile->ext = malloc(sizeof(double)*(profile->nmax+1));
+	if(description->profile->ext == NULL){
+		printf("Could not allocate description->profile->ext memory.\n");
+		exit(1);
+	}
+	
+	// copy description->profile values to profile
+	for(i=0;i<=profile->nmax;i++){
+		description->profile->z[i] = profile->z[i];
+		description->profile->cap[i] = profile->cap[i];
+		description->profile->ext[i] = profile->ext[i];
+	}
+
+	// Calculate open area
+	description->open_area = (description->profile->cap[0]/description->profile->ext[0]) * (description->profile->cap[0]/description->profile->ext[0]) * description->n_cap;
+
+	return description;
+}
+
+//===========================================
+// get the polycap_profile from a polycap_description
+polycap_profile* polycap_description_get_profile(polycap_description *description)
+{
```

Comment:
In plaats van een ganse nieuwe profile kopie te maken, waarom niet gewoon description->profile returnen?
Dikwijls wordt in zo een geval `const` gebruikt om het de gebruiker duidelijk te maken dat hij er niet mag mee gaan foefelen: de compiler zal een error geven normaal gezien (tenzij er een cast gebruikt wordt, maar dat is niet slim). De bedoeling is dat de documentatie informatie hierover zal bevatten: moet de return value ge'freed worden door de gebruiker of is de library de 'owner' van de pointer...

Reply:
Mja, een beetje gelijkaardig aan onze vorige discussie omtrent het maken van description utigaande van eerder gedefinieerde profile he. Een volledige kopie leek me veiliger indien de user inderdaad description al zou free'en, waardoor ook de inhoud in profile zou verdwijnen.

Op jouw manier, met die const, werkt het dan zo:
`const polycap_profile *profile;`
`profile = description->profile;`
`return profile;`
Of moet die const elders staan?

Reply:
Mijn voorstel is de functie te veranderen naar:

````c
const polycap_profile* polycap_description_get_profile(polycap_description *description) {
  return description->profile; // als de compiler hierover klaagt, doe dan een cast naar `const polycap_profile *`
}
````

Eigenlijk moet je je niet te veel aantrekken dat de gebruiker de boel naar de zak helpt door met de structs te gaan foefelen waarvan jij enkel een pointer hebt: zolang jij goed documenteert wat de gebruiker mag free'en en wat niet, is het de gebruiker zijn verantwoordelijkheid dat hij de library goed gebruikt, niet de jouwe. Vergeet ook niet dat ge aan het programmeren zijt in C, waar de minste fout meestal wordt afgestraft met een segfault: C programmeurs weten dat en zullen daarom de documentatie van een bepaalde library goed volgen.